### PR TITLE
community: added support for Guardrails for Amazon Bedrock

### PR DESCRIPTION
**Pull request by harelix**
Added support for optionally supplying 'Guardrails for Amazon Bedrock' on both types of model invocations (batch/regular and streaming) and for all models supported by the Amazon Bedrock service.

@baskaryan  @hwchase17

```python 
llm = Bedrock(model_id="<model_id>", client=bedrock,
                  model_kwargs={},
                  guardrails={"id": " <guardrail_id>",
                              "version": "<guardrail_version>",
                               "trace": True}, callbacks=[BedrockAsyncCallbackHandler()])

class BedrockAsyncCallbackHandler(AsyncCallbackHandler):
    """Async callback handler that can be used to handle callbacks from langchain."""

    async def on_llm_error(
            self,
            error: BaseException,
            **kwargs: Any,
    ) -> Any:
        reason = kwargs.get("reason")
        if reason == "GUARDRAIL_INTERVENED":
           # kwargs contains additional trace information sent by 'Guardrails for Bedrock' service.
            print(f"""Guardrails: {kwargs}""")


# streaming 
llm = Bedrock(model_id="<model_id>", client=bedrock,
                  model_kwargs={},
                  streaming=True,
                  guardrails={"id": "<guardrail_id>",
                              "version": "<guardrail_version>"})
```

### CODE REVIEW

File: libs/community/langchain_community/llms/bedrock.py [hunk 1]
```diff
@@ -191,6 +191,11 @@ class BedrockBase(BaseModel, ABC):
         "cohere": "stop_sequences",
     }
 
+    guardrails: Optional[Mapping[str, str]] = {
```
Comment (gilinachum) [inline libs/community/langchain_community/llms/bedrock.py:194]: I would add a string doc on what this property represents (as it's not trivial)

**Comment by gilinachum** [comment]
LGTM. Suggested to add a Docstring for the new Guardrails property.

File: libs/community/langchain_community/llms/bedrock.py (hunk 1 above)
Comment (harelix) [reply libs/community/langchain_community/llms/bedrock.py:194]: Thank you for pointing that out, @gilinachum. Recognizing the complexity of the property, I've updated it with a detailed docstring for better clarity on its purpose and usage.

**Review by harelix — COMMENTED** [review]
I've updated it with a detailed docstring for better clarity on its purpose and usage.

**Comment by Tom-HA** [comment]
+1

**Comment by guya-pecan** [comment]
+1

**Comment by romga** [comment]
👁️

**Review by gilinachum — APPROVED** [review]

File: libs/community/langchain_community/llms/bedrock.py [hunk 2]
```diff
@@ -191,6 +201,52 @@ class BedrockBase(BaseModel, ABC):
...
+    class BedrockAsyncCallbackHandler(AsyncCallbackHandler):
+        async def on_llm_error(
+            self,
+            error: BaseException,
+            **kwargs: Any,
+        ) -> Any:
+            reason = kwargs.get("reason")
+            if reason == "GUARDRAIL_INTERVENED":
+                ...Logic to handle guardrail intervention...
+    """
```
Comment (harelix) [inline libs/community/langchain_community/llms/bedrock.py:248]: Guardrails for Bedrock Intervention Callback Handler

**Comment by hwchase17** [comment]
@harelix  and @gilinachum i havent had time to test this, but am i correct in assuming that you both are from amazon? and i should feel confident just merging this in and trusting your testing/code review?

**Comment by harelix** [comment]
@hwchase17 true, we are both from Amazon, and we run our tests to verify that this is actually working and not introducing breaking changes.

**Review by 3coins — COMMENTED** [review]
@harelix
Thanks for adding this. There are some minor comments, and naming suggestions. Are you planning to add any unit tests? 
Here is a good PR to review for unit tests.
https://github.com/langchain-ai/langchain/pull/12477

File: libs/community/langchain_community/llms/bedrock.py [hunk 3]
```diff
@@ -254,6 +312,43 @@ def _get_provider(self) -> str:
...
+        """
+        try:
+
+            return isinstance(self.guardrails, dict) and bool(self.guardrails["id"]) and bool(self.guardrails["version"])
+
+        except KeyError as e:
+            raise TypeError("Guardrails must be a dictionary with 'id' and 'version' keys.") from e
+
+
+    def _get_guardrails_canonical_form(self) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (harelix) [reply libs/community/langchain_community/llms/bedrock.py:335]: done, please review.

File: libs/community/langchain_community/llms/bedrock.py [hunk 4]
```diff
@@ -284,8 +391,58 @@ def _prepare_input_and_invoke(
...
+           body dict():
+
+        Returns:
+           str: The text that was generated by the model.
+        """
+
+        if body.get(GUARDRAILS_BODY_KEY) == "GUARDRAIL_INTERVENED":
+            return True
+
+        return False
```
[the remark this answers is not in the document]
Comment (harelix) [reply libs/community/langchain_community/llms/bedrock.py]: done, please review

File: libs/community/langchain_community/llms/bedrock.py [hunk 5]
```diff
@@ -284,8 +391,58 @@ def _prepare_input_and_invoke(
...
+            during the request-response cycle with a Large Language Model (LLM).
+            """
+
+        if self._guardrails_enabled and self.guardrails.get("trace"):
+            if self._check_if_guardrails_intervened(body):
+                return {
+                    "signal" : True,
+                    "reason" : "GUARDRAIL_INTERVENED",
+                    "trace" : body.get(AMAZON_BEDROCK_TRACE_KEY)
+                }
```
[the remark this answers is not in the document]
Comment (harelix) [reply libs/community/langchain_community/llms/bedrock.py]: done, please review

File: libs/community/langchain_community/llms/bedrock.py [hunk 6]
```diff
@@ -284,8 +391,58 @@ def _prepare_input_and_invoke(
...
+
+    def _check_if_guardrails_intervened(self, body: dict) -> bool:
+        """
+        Determines if guardrails for Bedrock service intervened in the generation of response.
+
+        Args:
+           body dict():
+
+        Returns:
+           str: The text that was generated by the model.
```
[the remark this answers is not in the document]
Comment (harelix) [reply libs/community/langchain_community/llms/bedrock.py]: done, please review

File: libs/community/langchain_community/llms/bedrock.py [hunk 7]
```diff
@@ -284,8 +391,58 @@ def _prepare_input_and_invoke(
...
+        return {
+            "signal" : False,
+            "reason" : None,
+            "trace" : None,
+        }
+
+
+
+
+    def _check_if_guardrails_intervened(self, body: dict) -> bool:
```
[the remark this answers is not in the document]
Comment (harelix) [reply libs/community/langchain_community/llms/bedrock.py]: done, please review

File: libs/community/langchain_community/llms/bedrock.py [hunk 8]
```diff
@@ -284,8 +391,58 @@ def _prepare_input_and_invoke(
...
+        services_trace = self._inspect_bedrock_services_signal(body)
+
+        if services_trace.get("signal") and run_manager is not None:
+            run_manager.on_llm_error(Exception(f"Error raised by bedrock service: {services_trace.get('reason')}"),
+                                     **services_trace)
+
         return text
 
+
+    def _inspect_bedrock_services_signal(self, body: dict) -> dict:
```
[the remark this answers is not in the document]
Comment (harelix) [reply libs/community/langchain_community/llms/bedrock.py]: This method was originally composed as a general-purpose signal inspection, so I have changed it to: `_get_bedrock_services_signal`

**Comment by harelix** [comment]
> @harelix Thanks for adding this. There are some minor comments, and naming suggestions. Are you planning to add any unit tests? Here is a good PR to review for unit tests. #12477

I will include my integration tests in this PR

**Comment by harelix** [comment]
> @hwchase17 true, we are both from Amazon, and we run our tests to verify that this is actually working and not introducing breaking changes.

@hwchase17 - I've addressed the feedback and tested the changes in my pull request, and It's ready to be merged. Would you mind taking a look and merging it when you have a moment? I'd really appreciate it!

**Review by 3coins — APPROVED** [review]
@harelix 
Thanks for working on these changes. LGTM!

**Review by gilinachum — APPROVED** [review]
LGTM

**Comment by baskaryan** [comment]
apologies for the delay! code looks good but looks like there's now some merge conflicts. happy to merge once those are resolved

**Review by tom-ha — APPROVED** [review]

**Review by tom-ha — APPROVED** [review]

**Comment by harelix** [comment]
>  code looks good but looks like there's now some merge conflicts. happy to merge once those are resolved

@baskaryan;  resolved - thank you!

**Pull request merged**